### ISSUE TITLE
CPU topology option.

### ISSUE THREAD

**Issue by spikefishjohn** [GNS3/gns3-gui#3047]
TL;DR add a socket counter next to cpu and adjust arguments passed to qemu for the -smp argument.

Windows 10 has some built in limits on CPU topology. For example home edition only supports a single socket (not talking cores). 10 Pro supports 2 sockets and Pro Workstation Enterprise supports 4 sockets. Note: I don't think server has this limitation.

[https://en.wikipedia.org/wiki/Windows_10_editions](url)  - See Maximum CPU sockets

It seems like GNS3 is just using -smp $NUM as an argument to pass to QEMU. So with $NUM set to four QEMU is creating a motherboard with 4 cpu sockets that have a single core in each. QEMU supports passing sockets and cores via the -smp argument (as well as threads but thats beside the point).

I've tried using the advanced option on a VM to pass custom -smp and args but this doesn't take. I'm assuming because the -smp $NUM is passed much earlier in the command line.

**Comment by grossmj** [comment]
> I've tried using the advanced option on a VM to pass custom -smp and args but this doesn't take. I'm assuming because the -smp $NUM is passed much earlier in the command line.

This is strange because `additional settings` are passed last to the Qemu command line. I tested by adding `-smp cpus=4` and checking there are 4 cpus seen in the VM instead of 2.

![Screenshot from 2020-09-12 23-36-59](https://user-images.githubusercontent.com/4752614/92997274-3ba44980-f551-11ea-9024-7b888431bde3.png)

![Screenshot from 2020-09-12 23-36-50](https://user-images.githubusercontent.com/4752614/92997278-3f37d080-f551-11ea-985e-f810f135f179.png)

![Screenshot from 2020-09-12 23-57-22](https://user-images.githubusercontent.com/4752614/92997614-bbcbae80-f553-11ea-8ffc-d5cc55473cc5.png)

**Comment by b-ehlers** [comment]
>> I've tried using the advanced option on a VM to pass custom -smp and args but this doesn't take. I'm assuming because the -smp $NUM is passed much earlier in the command line.

> This is strange because additional settings are passed last to the Qemu command line. ...

Same here, again tested with Linux. I set the vCPUs to 8, which create a QEMU command line of `-smp cpus=8,maxcpus=8`.

Then I made three tests:
1. no additional options
2. additional option `-smp sockets=1`
3. additional option `-smp sockets=2`

Linux shows (with `cat /proc/cpuinfo`) for each virtual processor the physical id = socket number and the core id. Here the  physical id / core id for each processor on all three configurations:

| processor |  no additional options | socket=1  | socket=2  |
|---|---|---|---|
| 0 | 0 / 0 | 0 / 0 | 0 / 0 |
| 1 | 1 / 0 | 0 / 1 | 0 / 1 |
| 2 | 2 / 0 | 0 / 2 | 0 / 2 |
| 3 | 3 / 0 | 0 / 3 | 0 / 3 |
| 4 | 4 / 0 | 0 / 4 | 1 / 0 |
| 5 | 5 / 0 | 0 / 5 | 1 / 1 |
| 6 | 6 / 0 | 0 / 6 | 1 / 2 |
| 7 | 7 / 0 | 0 / 7 | 1 / 3 |

It indeed shows, that without any additional options GNS3/QEMU creates one socket per cpu, the core id is always 0. With the additional option `-smp sockets=1` QEMU uses only the one socket 0, the processors are put into cores 0 - 7. Likewise with `-smp sockets=2` QEMU uses two sockets (0 and 1) with cores 0 - 3.

So at least with a Linux VM I could easily configure the number of sockets with the additional option  
`-smp sockets=<number of sockets>`

**Comment by b-ehlers** [comment]
Additional tests with the Windows appliance / MSEdge on Windows 10 Enterprise 1809

General settings: vCPUs: 4

no additional options: 4 Proccessors with 1 Core each
-smp sockets=1: 1 Processor with 4 Cores
-smp sockets=2: 2 Processors with 2 Cores each

Maybe GNS3 should add the `-smp sockets=1` by default. That shouldn't make any difference on linux based VMs and would allow Windows users to use multiple cores without any additional configuration.

**Comment by grossmj** [comment]
> Maybe GNS3 should add the -smp sockets=1 by default. That shouldn't make any difference on linux based VMs and would allow Windows users to use multiple cores without any additional configuration.

Done :+1:

**Comment by spikefishjohn** [comment]
I don't have any idea what impact changing the default sockets option has in qemu performance wise (if any). I'll just say this seems to only effect windows 10 (and off shots)  and I seem the be the only one that every brought this up. I'm mean its cool this is fixed I'm just wondering what impact there is with making that change.

Side note, I know have a 128 CPU windows 10 VM that makes my laptop transform into a grilling station.
